### PR TITLE
Drop Hibernate from server

### PR DESCRIPTION
**Pull request by simonbrandhof**
Refactoring the registration of metrics at startup allows to drop Hibernate from server. 
Usage from batch will be automatically removed with drop of db connection.

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/metric/DefaultMetricFinder.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+import org.sonar.api.measures.Metric;
+import org.sonar.api.measures.MetricFinder;
+import org.sonar.core.metric.db.MetricDto;
+import org.sonar.core.persistence.DbSession;
+import org.sonar.core.persistence.MyBatis;
+import org.sonar.server.db.DbClient;
+
+import static com.google.common.collect.FluentIterable.from;
+
+public class DefaultMetricFinder implements MetricFinder {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/metric/DefaultMetricFinder.java]: MAJOR "MetricFinder" is deprecated, implement the suggested replacement instead. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1724)

File: server/sonar-server/src/main/java/org/sonar/server/metric/DefaultMetricFinder.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+      return dto.isEnabled();
+    }
+  }
+
+  private enum ToMetric implements Function<MetricDto, Metric> {
+    INSTANCE;
+
+    @Override
+    public Metric apply(@Nonnull MetricDto dto) {
+      Metric<Serializable> metric = new Metric<>();
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/metric/DefaultMetricFinder.java]: MINOR Constructor 'Metric(...)' is deprecated. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 1 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Comment by simonbrandhof** [comment]
SQ issues are expected. No need to fix.

File: server/sonar-server/src/main/java/org/sonar/server/startup/RegisterMetrics.java [hunk 3]
```diff
@@ -20,99 +20,142 @@
...
     for (Metrics metrics : metricsRepositories) {
       checkMetrics(metricsByRepository, metrics);
-      metricsToRegister.addAll(removeExistingUserManagedMetrics(metrics.getMetrics()));
+      metricsToRegister.addAll(metrics.getMetrics());
     }
 
     return metricsToRegister;
   }
 
-  private List<Metric> removeExistingUserManagedMetrics(List<Metric> metrics) {
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/startup/RegisterMetrics.java]: Why did you remove this check ?

**Comment by julienlancelot** [comment]
Just one question, and LGTM !

File: server/sonar-server/src/main/java/org/sonar/server/startup/RegisterMetrics.java (hunk 3 above)
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/startup/RegisterMetrics.java]: It was not removed but refactored. A custom metric is never updated when it already exists. This is implemented in line 96.

Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/startup/RegisterMetrics.java]: Ok !
